Tests/ApolloTests/Cache/ReadWriteFromStoreTests.swift

```diff
@@ -609,4 +609,155 @@ class ReadWriteFromStoreTests: XCTestCase, CacheDependentTesting, StoreLoading {
+        XCTAssertSuccessResult(jediResult)
+
+        self.store.withinReadTransaction({ transaction in
+          let query = HeroAndFriendsNamesWithIDsQuery(episode: .empire)
+          let data = try transaction.read(query: query)
+
+          XCTAssertEqual(data.hero?.__typename, "Droid")
+          XCTAssertEqual(data.hero?.name, "R2-D2")
+
+          let friendsNames = data.hero?.friends?.compactMap { $0?.name }
+          XCTAssertEqual(friendsNames, ["Luke Skywalker", "Leia Organa"])
+
+        }, completion: { empireResult in
+          defer { readCompletedExpectation.fulfill() }
+
+          XCTAssertSuccessResult(empireResult)
+        })
+      })
+    })
+
+    self.waitForExpectations(timeout: Self.defaultWaitTimeout)
+
+    //
+    // 3. Remove all objects matching the pattern `100`
+    // - This will remove `1002` (Han Solo, hero for the .newhope episode)
+    // - This will remove `1003` (Leia Organa, friend of the hero in .empire episode)
+    //
+
+    let removeCompletedExpectation = expectation(description: "Hero objects removed from cache by pattern")
+
+    store.withinReadWriteTransaction({ transaction in
+      try transaction.removeObjects(matching: "100")
+    }, completion: { result in
+      defer { removeCompletedExpectation.fulfill() }
+
+      XCTAssertSuccessResult(result)
+    })
+
+    self.waitForExpectations(timeout: Self.defaultWaitTimeout)
+
+    //
+    // 4. Attempt to read records after pattern removal
+    // - .newhope episode query expected to FAIL on the `hero` path
+    // - .jedi episdoe query expected to SUCCEED
+    // - .empire episode query expected to FAIL on the `hero.friends` path
+    //
+
+    let readAfterRemoveCompletedExpectation = expectation(description: "Read removed objects completed")
+
+    store.withinReadTransaction({ transaction in
+      let query = HeroAndFriendsNamesWithIDsQuery(episode: .newhope)
+      _ = try transaction.read(query: query)
+
+    }, completion: { newHopeResult in
+      XCTAssertFailureResult(newHopeResult) { error in
+        if let error = error as? GraphQLResultError {
+          XCTAssertEqual(error.path, ["hero"])
+          XCTAssertMatch(error.underlying, JSONDecodingError.missingValue)
+        } else {
+          XCTFail("Unexpected error: \(error)")
+        }
+      }
+
+      self.store.withinReadTransaction({ transaction in
+        let query = HeroAndFriendsNamesWithIDsQuery(episode: .jedi)
+        let data = try transaction.read(query: query)
+
+        XCTAssertEqual(data.hero?.__typename, "Human")
+        XCTAssertEqual(data.hero?.name, "Luke Skywalker")
+
+      }, completion: { jediResult in
+        XCTAssertSuccessResult(jediResult)
+
+        self.store.withinReadTransaction({ transaction in
+          let query = HeroAndFriendsNamesWithIDsQuery(episode: .empire)
+          _ = try transaction.read(query: query)
+
+        }, completion: { empireResult in
+          defer { readAfterRemoveCompletedExpectation.fulfill() }
```

Comment:
Instead of having one expectation and three nested reads, it's probably better here to have three separate reads with three separate expectations. a) Way easier to read b) Can execute concurrently

[same thought for step 2 as well]

Reply:
^still would love to see this

Reply:
This is done in 9cbd23ab